lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h

```diff
@@ -423,20 +428,65 @@ class SwiftASTContext : public TypeSystemSwift {
 
   CompilerType GetErrorType() override;
 
-  bool HasErrors();
+  /// Error handling
+  /// \{
+  bool HasDiagnostics() const;
+  bool HasClangImporterErrors() const;
 
-  // NEVER call this without checking HasFatalErrors() first.
-  // This clears the fatal-error state which is terrible.
-  // We will assert if you clear an actual fatal error.
-  void ClearDiagnostics();
+  void AddDiagnostic(DiagnosticSeverity severity, llvm::StringRef message);
+  void RaiseFatalError(std::string msg) const {
+    m_fatal_errors.SetErrorString(msg);
+  }
+  static bool HasFatalErrors(swift::ASTContext *ast_context);
+  bool HasFatalErrors() const {
+    return m_fatal_errors.Fail() || HasFatalErrors(m_ast_context_ap.get());
+  }
 
+  /// Return only fatal errors.
+  Status GetFatalErrors() const;
+  /// Notify the Process about any Swift or ClangImporter errors.
+  void DiagnoseWarnings(Process &process, Module &module) const override;
+  
   bool SetColorizeDiagnostics(bool b);
-  void AddErrorStatusAsGenericDiagnostic(Status error);
 
   void PrintDiagnostics(DiagnosticManager &diagnostic_manager,
                         uint32_t bufferID = UINT32_MAX, uint32_t first_line = 0,
                         uint32_t last_line = UINT32_MAX) const;
 
+  /// A set of indices into the diagnostic vectors to mark the start
+  /// of a transaction.
+  struct DiagnosticCursor {
+    size_t swift = 0;
+    size_t clang = 0;
+    size_t lldb = 0;
+    size_t m_num_swift_errors = 0;
+  };
+
+  /// A lightweigt abstraction that stits on top of a diagnostic
+  /// consumer that can be used capture diagnostics for one transation
+  /// and restore the state of the consumer afterwards.
+  class TransactionDiagnostics {
+    swift::DiagnosticConsumer &m_consumer;
+    const DiagnosticCursor m_cursor;
+
+  public:
+    TransactionDiagnostics(swift::DiagnosticConsumer &consumer);
+    ~TransactionDiagnostics();
+    /// Print all diagnostics that happened during the lifetime of
+    /// this object to diagnostic_manager. If none is found, print the
+    /// persistent diagnostics form the parent consumer.
+    void PrintDiagnostics(DiagnosticManager &diagnostic_manager,
+                          uint32_t bufferID = UINT32_MAX,
+                          uint32_t first_line = 0,
+                          uint32_t last_line = UINT32_MAX) const;
+    bool HasErrors() const;
+    /// Return all errors and warnings that happened during the lifetime of this
+    /// object.
+    Status GetAllErrors() const;
+  };
+  std::unique_ptr<TransactionDiagnostics> forkDiagnosticConsumer();
```

Comment:
Because it's an overlay to the main diagnostic consumer. All the diagnostics are still handled by the parent consumer, but it deletes certain kinds of diagnostics after destruction.